### PR TITLE
Add support for Razer Cynosa V2

### LINKED ISSUE #1240: Support for Razer Cynosa V2
### Device Info

* Name: Razer Cynosa V2
* Product Number: RZ03-0340
* Model Number: RZ03-03400200-R3U1

### Device Information
```
Bus 001 Device 002: ID 1532:025e Razer USA, Ltd Razer Cynosa V2
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               2.00
  bDeviceClass            0 
  bDeviceSubClass         0 
  bDeviceProtocol         0 
  bMaxPacketSize0        64
  idVendor           0x1532 Razer USA, Ltd
  idProduct          0x025e 
  bcdDevice            2.00
  iManufacturer           1 Razer
  iProduct                2 Razer Cynosa V2
  iSerial                 0 
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x0054
    bNumInterfaces          3
    bConfigurationValue     1
    iConfiguration          0 
    bmAttributes         0xa0
      (Bus Powered)
      Remote Wakeup
    MaxPower              500mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
```

```
total 0
drwxr-xr-x 2 root root 240 2020-08-29 12:02 ./
drwxr-xr-x 4 root root 680 2020-08-29 12:02 ../
lrwxrwxrwx 1 root root  10 2020-08-29 11:40 usb-Burr-Brown_from_TI_USB_Audio_DAC-event-if02 -> ../event11
lrwxrwxrwx 1 root root  10 2020-08-29 11:40 usb-Logitech_G_Pro_Gaming_Mouse_108B385C3437-event-if01 -> ../event12
lrwxrwxrwx 1 root root   9 2020-08-29 11:40 usb-Logitech_G_Pro_Gaming_Mouse_108B385C3437-event-mouse -> ../event9
lrwxrwxrwx 1 root root  10 2020-08-29 11:40 usb-Logitech_G_Pro_Gaming_Mouse_108B385C3437-if01-event-kbd -> ../event10
lrwxrwxrwx 1 root root   9 2020-08-29 11:40 usb-Logitech_G_Pro_Gaming_Mouse_108B385C3437-mouse -> ../mouse1
lrwxrwxrwx 1 root root   9 2020-08-29 11:40 usb-Razer_Razer_Cynosa_V2-event-if01 -> ../event6
lrwxrwxrwx 1 root root   9 2020-08-29 11:40 usb-Razer_Razer_Cynosa_V2-event-kbd -> ../event2
lrwxrwxrwx 1 root root   9 2020-08-29 11:40 usb-Razer_Razer_Cynosa_V2-if01-event-kbd -> ../event3
lrwxrwxrwx 1 root root   9 2020-08-29 11:40 usb-Razer_Razer_Cynosa_V2-if02-event-mouse -> ../event7
lrwxrwxrwx 1 root root   9 2020-08-29 11:40 usb-Razer_Razer_Cynosa_V2-if02-mouse -> ../mouse0
```

### PR DESCRIPTION
**Pull request by jcanso**
Hopefully I managed to tick all the boxes.

Fixes #1240

### CODE REVIEW

**Review by nightsky30 — COMMENTED** [review]
@jcanso I'm not sure if the fake drivers are required. I've tried to include them with my device PRs.

I would definitely add an entry for this device in README.md.

Also, add a case statement for this device for removal of device files similar to what you did with creating device files:
https://github.com/openrazer/openrazer/blob/e1e09e619faad8f1cbea7a9f948b1acc07fa8630/driver/razerkbd_driver.c#L2688-L2694

Thanks for contributing this :)

File: daemon/openrazer_daemon/hardware/keyboards.py [hunk 1]
```diff
@@ -699,6 +699,27 @@ class RazerCynosaChroma(_RippleKeyboard):
...
+    WAVE_DIRS = (0, 1)
+    MATRIX_DIMS = [6, 22]
+    METHODS = ['get_device_type_keyboard', 'set_wave_effect', 'set_static_effect', 'set_spectrum_effect',
+               'set_reactive_effect', 'set_none_effect', 'set_breath_random_effect', 'set_breath_single_effect', 'set_breath_dual_effect',
+               'set_custom_effect', 'set_key_row', 'get_game_mode', 'set_game_mode', 'get_macro_mode', 'set_macro_mode',
+               'get_macro_effect', 'set_macro_effect', 'get_macros', 'delete_macro', 'add_macro',
+               'set_starlight_random_effect', 'set_starlight_single_effect', 'set_starlight_dual_effect',
+               'set_ripple_effect', 'set_ripple_effect_random_colour']
+
+    DEVICE_IMAGE = "https://assets.razerzone.com/eeimages/support/products/1256/1256_cynosa_chroma.png"
```
Comment (nightsky30) [inline daemon/openrazer_daemon/hardware/keyboards.py]: It looks like the device image could be updated here: https://assets.razerzone.com/eeimages/support/products/1694/1694-2.png

Comment (jcanso) [reply daemon/openrazer_daemon/hardware/keyboards.py]: thanks for the suggestions!

File: README.md [hunk 2]
```diff
@@ -60,6 +60,7 @@ The below devices are fully feature supported by OpenRazer which means all avail
 | Razer Blade 15 Studio Edition (2019)          |  1532:024D  |
 | Razer Blade Stealth (Early 2020)              |  1532:0252  |
 | Razer Blade 15 Advanced (2020)                |  1532:0253  |
+| Razer Cynosa Chroma V2                        |  1532:025E  |
```
Comment (z3ntu) [inline README.md]: Isn't this device just called `Razer Cynosa V2`?

**Comment by z3ntu** [comment]
Do all effects specified here work on the device?

File: daemon/openrazer_daemon/hardware/keyboards.py [hunk 3]
```diff
@@ -699,6 +699,27 @@ class RazerCynosaChroma(_RippleKeyboard):
     DEVICE_IMAGE = "https://assets.razerzone.com/eeimages/support/products/1256/1256_cynosa_chroma.png"
 
 
+class RazerCynosaChromaV2(_RippleKeyboard):
+    """
+    Class for the Razer Cynosa Chroma V2
```
Comment (nightsky30) [inline daemon/openrazer_daemon/hardware/keyboards.py]: Based on z3ntu's [comment](https://github.com/openrazer/openrazer/pull/1257/files#r495458959) I would think it needs to be updated here (class name and comment).

File: daemon/openrazer_daemon/hardware/keyboards.py [hunk 4]
```diff
@@ -699,6 +699,27 @@ class RazerCynosaChroma(_RippleKeyboard):
...
+    """
+    EVENT_FILE_REGEX = re.compile(r'.*Cynosa_Chroma_V2(-if01)?-event-kbd')
```
Comment (nightsky30) [inline daemon/openrazer_daemon/hardware/keyboards.py]: Also, the regex used here won't match the event file(s) per output in #1240:

```
lrwxrwxrwx 1 root root   9 2020-08-29 11:40 usb-Razer_Razer_Cynosa_V2-event-if01 -> ../event6
lrwxrwxrwx 1 root root   9 2020-08-29 11:40 usb-Razer_Razer_Cynosa_V2-event-kbd -> ../event2
lrwxrwxrwx 1 root root   9 2020-08-29 11:40 usb-Razer_Razer_Cynosa_V2-if01-event-kbd -> ../event3
lrwxrwxrwx 1 root root   9 2020-08-29 11:40 usb-Razer_Razer_Cynosa_V2-if02-event-mouse -> ../event7
lrwxrwxrwx 1 root root   9 2020-08-29 11:40 usb-Razer_Razer_Cynosa_V2-if02-mouse -> ../mouse0
```

File: driver/razerkbd_driver.c [hunk 5]
```diff
@@ -546,6 +546,11 @@ static ssize_t razer_attr_read_device_type(struct device *dev, struct device_att
     case USB_DEVICE_ID_RAZER_ANANSI:
         device_type = "Razer Anansi\n";
         break;
+
+    case USB_DEVICE_ID_RAZER_CYNOSA_CHROMA_V2:
```
Comment (nightsky30) [inline driver/razerkbd_driver.c]: Update device ID without "CHROMA" throughout driver.

File: driver/razerkbd_driver.h [hunk 6]
```diff
@@ -68,6 +68,7 @@
 #define USB_DEVICE_ID_RAZER_BLADE_STUDIO_EDITION_2019 0x024d
 #define USB_DEVICE_ID_RAZER_BLADE_STEALTH_EARLY_2020 0x0252
 #define USB_DEVICE_ID_RAZER_BLADE_15_ADV_2020 0x0253
+#define USB_DEVICE_ID_RAZER_CYNOSA_CHROMA_V2 0x025E
```
Comment (nightsky30) [inline driver/razerkbd_driver.h]: Same here :)

**Review by nightsky30 — CHANGES_REQUESTED** [review]

File: pylib/openrazer/_fake_driver/razercynosachromav2.cfg [hunk 7]
```diff
@@ -0,0 +1,23 @@
+[device]
+dir_name = 0003:1532:025E.0001
+name = Razer Cynosa Chroma V2
+event = Cynosa_Chroma_V2-event-kbd
+        Cynosa_Chroma_V2-if01-event-kbd
```
Comment (nightsky30) [inline pylib/openrazer/_fake_driver/razercynosachromav2.cfg]: Same here for name and event regex :)

**Review by nightsky30 — CHANGES_REQUESTED** [review]

**Review by nightsky30 — APPROVED** [review]

**Comment by jcanso** [comment]
> Do all effects specified here work on the device?

I checked all of the ones in the polychromatic systray thing and all of the effects worked

**Comment by evgengal** [comment]
Please merge! Can't wait )

**Review by z3ntu — APPROVED** [review]
LGTM, thanks!

**Comment by MCSH** [comment]
Not sure if this is the correct place to mention this (will open an issue when I have enough time with the correct info, logs, etc) but the volume up and volume down keys work as a down arrow and a left arrow respectively whenever the daemon is running.

**Comment by z3ntu** [comment]
@MCSH Please open a new issue for that otherwise it'll get lost.

**Pull request merged**